azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,405 @@
+        self._fabric_changed_detail = None
+        self._reason = None
+        self._reset_if_necessary()
+
+    def get_ext_config(self, incarnation, ext_conf_uri):
+        # If we don't have a uri, return an empty extensions config
+        is_startup = False
+        if ext_conf_uri is None:
+            return GenericExtensionsConfig(ExtensionsConfig(None), False, self)
+
+        # Get the Fabric goal state and whether it changed
+        fabric_changed = self._get_fabric_changed(incarnation)
+
+        # Get the VmArtifactsProfile and whether fast track changed, if enabled
+        artifacts_profile, fast_track_changed = self._get_fast_track_details()
+
+        self._pending_mode = self._decide_what_to_process(fabric_changed, fast_track_changed)
+        if self._last_mode is None:
+            is_startup = True
+            logger.info("Using {0} for the first call to extensions. Reason={1}", self._pending_mode, self._reason)
+        elif self._pending_mode != self._last_mode:
+            logger.info("Processing from previous mode {0}. New mode is {1}. Reason={2}",
+                        self._last_mode, self._pending_mode, self._reason)
+
+        extensions_config = None
+        changed = False
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            xml_text = self._wire_client.fetch_config(ext_conf_uri, self._wire_client.get_header())
+            extensions_config = ExtensionsConfig(xml_text)
+            changed = fabric_changed or is_startup
+            if fast_track_changed:
+                # If FastTrack changed too, then save the artifacts profile because the next time
+                # we retrieve it, we'll receive a 304 because the etag didn't change
+                self._saved_artifacts_profile = artifacts_profile
+        else:
+            extensions_config = artifacts_profile.transform_to_extensions_config()
+            changed = fast_track_changed or is_startup
+            self._saved_artifacts_profile = artifacts_profile
+
+        if changed:
+            if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+                self._remove_extensions_if_necessary(extensions_config)
+                self._pending_fabric_incarnation = str(incarnation)
+            else:
+                self._pending_fast_track_seq_no = artifacts_profile.get_sequence_number()
+            logger.info("Handling extension updates. LastMode={0}, IsStartup={1}", self._last_mode, is_startup)
+
+        ext_conf = GenericExtensionsConfig(extensions_config, changed, self)
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            # We only need to retrieve certs if the Fabric incarnation changes. FastTrack won't change them
+            ext_conf.is_fabric_change = True
+        return ext_conf
+
+    def _get_fast_track_details(self):
+        artifacts_profile = None
+        fast_track_changed = False
+        
+        if conf.get_extensions_fast_track_enabled():
```

Comment:
didn't this kind of checks also required to check whether the host ga plugin supports fast track?

Reply:
We do check whether host ga plugin supports fast track, but not here.
CRP will only use FastTrack if GA says it support it.
GA advertises that it supports FastTrack through the SupportedFeatures dictionary in the status blob.
When GA is determining the SupportedFeatures, it takes into account whether HostGAPlugin has received an etag.
So we'll never receive a FastTrack GS unless HostGAPlugin supports it (and this check may be removed soon because HostGAPlugin has already been rolled out)